### PR TITLE
Return large buffers to the bufferPool #22723

### PR DESCRIPTION
**Pull request by astonchev**
In the large outbound flow EnvelopeBuffers aquired by Encoder must be
returned to the same buffer pool by the AeronSink. Otherwise one of
the following may happen:
* Full GC (System.gc())
* java.lang.OutOfMemoryError: Direct buffer memory
* kernel killing the process (OOM-killer)

see issue #22723

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @astonchev,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by astonchev** [comment]
Yes, I did sign the CLA.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 1]
```diff
@@ -977,6 +977,16 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
       envelopeBufferPool, giveUpAfter, createFlightRecorderEventSink()))
   }
 
+  private def aeronSink(outboundContext: OutboundContext, streamId: Int,
+                        bufferPool: EnvelopeBufferPool): Sink[EnvelopeBuffer, Future[Done]] = {
```
Comment (raboof) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: Perhaps the other `private def aeronSink` should now just call this one with `envelopeBufferPool`?

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 2]
```diff
@@ -963,7 +963,7 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
                                  bufferPool: EnvelopeBufferPool): Sink[OutboundEnvelope, (OutboundCompressionAccess, Future[Done])] = {
 
     outboundLane(outboundContext, bufferPool)
-      .toMat(aeronSink(outboundContext, streamId))(Keep.both)
+      .toMat(aeronSink(outboundContext, streamId, bufferPool))(Keep.both)
```
Comment (raboof) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: Good catch, thanks!

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/DirectMemorySpec.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+package akka.remote.artery
+
+import scala.concurrent.duration._
+import akka.actor.{ Actor, ActorPath, ActorRef, Props }
+import akka.remote.RemotingMultiNodeSpec
+import akka.remote.testkit.{ MultiNodeConfig, MultiNodeSpec, STMultiNodeSpec }
+import akka.testkit.ImplicitSender
+import com.typesafe.config.ConfigFactory
+
+object DirectMemorySpec extends MultiNodeConfig {
```
Comment (raboof) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/DirectMemorySpec.scala]: :+1:

**Review by raboof — APPROVED** [review]
LGTM nice 'minimal' test case, good observation!

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 1 above)
Comment (astonchev) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: yes, let's keep it Dry

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 4]
```diff
@@ -963,18 +963,23 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
...
   }
 
   def aeronSink(outboundContext: OutboundContext): Sink[EnvelopeBuffer, Future[Done]] =
     aeronSink(outboundContext, ordinaryStreamId)
 
-  private def aeronSink(outboundContext: OutboundContext, streamId: Int): Sink[EnvelopeBuffer, Future[Done]] = {
+  private def aeronSink(outboundContext: OutboundContext, streamId: Int): Sink[EnvelopeBuffer, Future[Done]] =
+    aeronSink(outboundContext, streamId, envelopeBufferPool)
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: I think it's better to remove this overload and change at the call site (to make the parameter an explicit choice)

**Review by patriknw — APPROVED** [review]
LGTM, very good fix and test. Optional adjustment.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 4 above)
Comment (astonchev) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: OK, thanks.

**Comment by astonchev** [comment]
PLS BUILD

**Comment by astonchev** [comment]
Can somebody please restart the build? It failed due to "sbt.ResolveException: download failed"

**Comment by raboof** [comment]
PLS BUILD

**Comment by patriknw** [comment]
@raboof Please backport (cherry-pick) this with a PR to `release-2.4` branch.

**Comment by raboof** [comment]
@patriknw like https://github.com/akka/akka/pull/22743 ?

**Comment by ktoso** [comment]
Yup, thanks :)

**Comment by patriknw** [comment]
thanks

**Comment by patriknw** [comment]
and to make sure that it shows up in some way in both 2.4.x and 2.5.x release notes (github milestones) I usually mark the issue with 2.4.17 and the master PR with 2.5.1.

**Comment by astonchev** [comment]
Thanks for the quick response. When should we expect the next release (approximately)?

**Comment by patriknw** [comment]
I think we'll release next week.

**Pull request merged**